File: Example/DeepleaperAdExample/DeepleaperAdExample/NativeViewController.m
//
//  NativeViewController.m
//  DeepleaperAdExample
//
//  Created by 刘畅 on 2017/6/22.
//  Copyright © 2017年 刘畅. All rights reserved.
//

#import "NativeViewController.h"
#import "FakeData.h"
#import "LandPageViewController.h"

@import DeepleaperAd;

@interface NativeViewController ()<DeepleaperFeedAdManagerDelegate>
@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property(nonatomic,copy) NSMutableArray* sourceArray;//数据数组
@property(nonatomic,assign) NSInteger indexRow;//上次点击的行
@property(nonatomic,strong) DeepleaperFeedAdManager* fam;
@end

@implementation NativeViewController

-(NSMutableArray*)sourceArray{
    if (!_sourceArray) {
        _sourceArray = [NSMutableArray new];
    }
    return _sourceArray;
}

- (void)viewDidLoad {
    [super viewDidLoad];

    // Do any additional setup after loading the view.
    for (NSUInteger i = 0 ; i<50; i++) {
        FakeData* data = [FakeData new];
        data.title = [NSString stringWithFormat:@"第%ld行",(unsigned long)i];
        [self.sourceArray addObject:data];
    }
}
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath{
    FakeData* data = _sourceArray[indexPath.row];
    //如果禁用了autoMode，则手动返回行高
    if (data.nativeAdView && !data.nativeAdView.autoMode) {
        return data.height;
    }
    else{
        return UITableViewAutomaticDimension;
    }
    
}
- (CGFloat)tableView:(UITableView *)tableView estimatedHeightForRowAtIndexPath:(NSIndexPath *)indexPath{
    return UITableViewAutomaticDimension;
}
- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _sourceArray.count;
}
- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    tableView.separatorStyle = UITableViewCellSeparatorStyleSingleLine;
    FakeData* data = _sourceArray[indexPath.row];
    if (data.title) {
        static NSString *ID = @"cell";
        UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:ID];
        if (cell == nil) {
            cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:ID];
        }
        cell.textLabel.text = data.title;
        return cell;
    }
    //广告
    else{
        static NSString *adID = @"adCell";
        UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:adID];
        if (cell == nil) {
            cell = [[UITableViewCell alloc]initWithStyle:UITableViewCellStyleDefault reuseIdentifier:adID];
        }
        for (UIView* view in cell.contentView.subviews) {
            [view removeFromSuperview];
        }
        [cell.contentView addSubview:data.nativeAdView];
        return cell;
    }
}
-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath{
    //记录点击行
    _indexRow = indexPath.row+1;
    //初始化
    _fam = [DeepleaperFeedAdManager getFeedAdManager];
    _fam.delegate = self;
    _fam.placementID = @"4cc2f95f-629f-446f-83cd-46ca8593f76b";
    _fam.sandBox = NO;
    NSString* urlString = @"http://m.d1cm.com/news/2017032089009.shtml"; //跳转的落地页url
    //请求广告
    [_fam loadAdWithFeedAdType:FEEDAD_NATIVE andUrl:urlString];
    //跳转落地页
    LandPageViewController* lvc = [LandPageViewController new];
    lvc.url = [NSURL URLWithString:urlString];
    [self presentViewController:lvc animated:YES completion:^{
    }];
}

//返回广告成功
-(void)didLoadNativeAdView:(DeepleaperNativeAdView *)nativeAdView{
//使用autoMode
    //加入到数据数组
    FakeData* data = [[FakeData alloc]init];
    data.nativeAdView = nativeAdView;
    [_sourceArray insertObject:data atIndex:_indexRow];
    dispatch_async(dispatch_get_main_queue(), ^(){
        //刷新
        [self.tableView reloadData];
    });
//不适用autoMode，手动布局
    /*
        FakeData* data = [[FakeData alloc]init];
        dispatch_async(dispatch_get_main_queue(), ^(){
     //禁用autoMode
            nativeAdView.autoMode = NO;
     //设置title
            nativeAdView.title ? [nativeAdView.title setFrame:CGRectMake(0, 0, 375, 20)] : nil;
     //根据不同物料类型设置位置
            if (nativeAdView.nativeType == NATIVE_BIGIMAGE || nativeAdView.nativeType == NATIVE_SINGLEIMAGE || nativeAdView.nativeType == NATIVE_BIGVIDEO || nativeAdView.nativeType == NATIVE_SINGLEVIDEO) {
                [nativeAdView.originalityView setFrame:CGRectMake(0, nativeAdView.title?nativeAdView.title.frame.size.height+10:0, 375, 375*nativeAdView.scale)];
                [nativeAdView.originalityView.firstImageView setFrame:nativeAdView.originalityView.bounds];
                if (nativeAdView.originalityView.PlayImageView) {
                    [nativeAdView.originalityView.PlayImageView setFrame:CGRectMake(0, 0, nativeAdView.originalityView.bounds.size.height/3, nativeAdView.originalityView.bounds.size.height/3)];
                    [nativeAdView.originalityView.PlayImageView setCenter:nativeAdView.originalityView.center];
                }
            }
            else if (nativeAdView.nativeType == NATIVE_THREEIMAGES){
                [nativeAdView.originalityView setFrame:CGRectMake(0, nativeAdView.title?nativeAdView.title.frame.size.height+10:0, 375, (375/3-10)*nativeAdView.scale)];
                [nativeAdView.originalityView.firstImageView setFrame:CGRectMake(0, 0, 375/3-10, (375/3-10)*nativeAdView.scale)];
                [nativeAdView.originalityView.secondImageView setFrame:CGRectMake(0, 0, 375/3-10, (375/3-10)*nativeAdView.scale)];
                [nativeAdView.originalityView.secondImageView setCenter:nativeAdView.originalityView.center];
                [nativeAdView.originalityView.thirdImageView setFrame:CGRectMake(375-(375/3-10), 0, 375/3-10, (375/3-10)*nativeAdView.scale)];
            }
            else if (nativeAdView.nativeType == NATIVE_SMALLIMAGE){
                [nativeAdView.originalityView setFrame:CGRectMake(0, nativeAdView.title?nativeAdView.title.frame.size.height+10:0, 375, (375/3-10)*nativeAdView.scale+20)];
                [nativeAdView.originalityView.firstImageView setFrame:CGRectMake(0, 10, 375/3-10, (375/3-10)*nativeAdView.scale)];
                [nativeAdView.originalityView.originalityText  setFrame:CGRectMake(375/3, 0, 375/3*2, (375/3-10)*nativeAdView.scale+20)];
            }
            else if (nativeAdView.nativeType == NATIVE_SINGLETITLE){
                [nativeAdView.originalityView setFrame:CGRectMake(0, nativeAdView.title?nativeAdView.title.frame.size.height+10:0, 375, (375/3-10)*nativeAdView.scale+20)];
                [nativeAdView.originalityView.originalityText  setFrame:CGRectMake(0, 0, 375, 100)];
            }
     //判断是否有附加创意
            if (nativeAdView.additionalType) {
                nativeAdView.additionalView.frame = CGRectMake(0, nativeAdView.originalityView.frame.origin.y+nativeAdView.originalityView.frame.size.height, 375, 30);
                nativeAdView.additionalView.additionalText.frame = CGRectMake(0, 0, 200, 30);
                [nativeAdView.additionalView.additionalButton sizeToFit];
                nativeAdView.additionalView.additionalButton.frame = CGRectMake(375-nativeAdView.additionalView.additionalButton.bounds.size.width-10, 0, nativeAdView.additionalView.additionalButton.bounds.size.width, nativeAdView.additionalView.additionalButton.bounds.size.height);
            }
            nativeAdView.logoView.frame = CGRectMake(0, nativeAdView.additionalType ? nativeAdView.additionalView.frame.origin.y+nativeAdView.additionalView.frame.size.height : nativeAdView.originalityView.frame.origin.y+nativeAdView.originalityView.frame.size.height, 375, 20);
            nativeAdView.logoView.logoImageView.frame = CGRectMake(5, 5, 10/0.0625, 10);
            nativeAdView.logoView.closeButton.frame = CGRectMake(375-20-10, 0, 20, 20);
            data.height = nativeAdView.logoView.frame.origin.y+nativeAdView.logoView.frame.size.height;
            data.nativeAdView = nativeAdView;
            [_sourceArray insertObject:data atIndex:_indexRow];
            [self.tableView reloadData];
        });
     */
}
//关闭了广告
-(void)didCloseNativeAdView:(DeepleaperNativeAdView *)nativeAdView{
    for (NSInteger i = _sourceArray.count-1 ; i>=0; i--) {
        FakeData* data = _sourceArray[i];
        if (data.nativeAdView.isClosed) {
            [_sourceArray removeObjectAtIndex:i];
            dispatch_async(dispatch_get_main_queue(), ^{
                [self.tableView deleteRowsAtIndexPaths:@[[NSIndexPath indexPathForRow:i inSection:0]]  withRowAnimation:UITableViewRowAnimationFade];
            });
            break;
        }
    }
}
//广告请求失败
-(void)didFailLoadNativeAdView:(NSError *)error{
    NSLog(@"error:%@",error);
}
- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
 #pragma mark - Navigation
 
 // In a storyboard-based application, you will often want to do a little preparation before navigation
 - (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
 // Get the new view controller using [segue destinationViewController].
 // Pass the selected object to the new view controller.
 }
 */

@end

File: Example/DeepleaperAdExample_Swift/DeepleaperAdExample_Swift/NativeViewController.swift
//
//  NativeViewController.swift
//  DeepleaperAdExample_Swift
//
//  Created by 刘畅 on 2017/7/14.
//  Copyright © 2017年 刘畅. All rights reserved.
//

import UIKit
import DeepleaperAd;
class NativeViewController: UIViewController,DeepleaperFeedAdManagerDelegate,UITableViewDelegate,UITableViewDataSource {
    @IBOutlet weak var tableView: UITableView!
    var _sourceArray:[FakeData] = Array();
    var _indexRow = 0;
    var _fam : DeepleaperFeedAdManager!;
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // 假数据数组
        for i in 0..<50{
            let data = FakeData();
            data.title = String.init(format: "第%d行",i);
            _sourceArray.append(data);
        }
    }
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return _sourceArray.count;
    }
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let data = _sourceArray[indexPath.row];
        if (data.nativeAdView != nil && !(data.nativeAdView?.autoMode)!){
            return data.height;
        }
        else{
            return UITableViewAutomaticDimension;
        }
    }
    func tableView(_ tableView: UITableView, estimatedHeightForRowAt indexPath: IndexPath) -> CGFloat {
        return UITableViewAutomaticDimension;
    }
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let data = _sourceArray[indexPath.row];
        if(data.title != nil){
            let ID = "cell";
            var cell = tableView.dequeueReusableCell(withIdentifier: ID);
            if(cell == nil){
                cell = UITableViewCell.init(style: .default, reuseIdentifier: ID);
            }
            cell?.textLabel?.text = data.title;
            return cell!;
        }
        else{
            let adID = "adCell";
            var cell = tableView.dequeueReusableCell(withIdentifier: adID);
            if(cell == nil){
                cell = UITableViewCell.init(style: .default, reuseIdentifier: adID);
            }
            for view in (cell?.contentView.subviews)! {
                view.removeFromSuperview();
            }
            cell?.contentView.addSubview(data.nativeAdView!);
            return cell!;
        }
    }
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        _indexRow = indexPath.row;
        //初始化
        _fam = DeepleaperFeedAdManager.getFeedAdManager();
        _fam.delegate = self;
        _fam.placementID = "4cc2f95f-629f-446f-83cd-46ca8593f76b";
        _fam.sandBox = false;
        let urlString = "http://m.d1cm.com/news/2017032089009.shtml";
        //请求原生广告
        _fam.loadAd(with: FEEDAD_NATIVE, andUrl: urlString);
        let lvc = LandPageViewController();
        lvc.url = URL.init(string: urlString);
        self.present(lvc, animated: true) {
        };
    }
    func didLoad(_ nativeAdView: DeepleaperNativeAdView) {
//        使用autoMode
//        加入到数据源
                let data = FakeData();
                data.nativeAdView = nativeAdView;
                _sourceArray.insert(data, at: _indexRow);
                DispatchQueue.main.async {
                    self.tableView.reloadData();
                }
        
        /*
        //不适用自动布局
        let data = FakeData();
        DispatchQueue.main.async {
            //设置详细参数
            nativeAdView.autoMode = false;
            if((nativeAdView.title) != nil){
                nativeAdView.title?.frame = CGRect.init(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: 20);
            }
            if(nativeAdView.nativeType == NATIVE_BIGIMAGE || nativeAdView.nativeType == NATIVE_SINGLEIMAGE || nativeAdView.nativeType == NATIVE_BIGVIDEO || nativeAdView.nativeType == NATIVE_SINGLEVIDEO){
                (nativeAdView.title == nil) ? (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.width*nativeAdView.scale)) : (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: (nativeAdView.title?.frame.size.height)!+10, width: UIScreen.main.bounds.size.width, height: UIScreen.main.bounds.size.width*nativeAdView.scale));
                nativeAdView.originalityView.firstImageView.frame = nativeAdView.originalityView.bounds;
                if((nativeAdView.originalityView.playImageView) != nil){
                    nativeAdView.originalityView.playImageView.frame = CGRect.init(x: 0, y: 0, width: nativeAdView.originalityView.bounds.size.height/3, height: nativeAdView.originalityView.bounds.size.height/3);
                    nativeAdView.originalityView.playImageView.center = nativeAdView.originalityView.center;
                }
            }
            else if(nativeAdView.nativeType == NATIVE_THREEIMAGES){
                (nativeAdView.title == nil) ? (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale)) : (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: (nativeAdView.title?.frame.size.height)!+10, width: UIScreen.main.bounds.size.width, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale));
                nativeAdView.originalityView.firstImageView.frame = CGRect.init(x: 0, y: 0, width: UIScreen.main.bounds.size.width/3-10, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale);
                nativeAdView.originalityView.secondImageView.frame = CGRect.init(x: UIScreen.main.bounds.size.width/3+5, y: 0, width: UIScreen.main.bounds.size.width/3-10, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale);
                //nativeAdView.originalityView.secondImageView.center = nativeAdView.originalityView.center;
                nativeAdView.originalityView.thirdImageView.frame = CGRect.init(x: UIScreen.main.bounds.size.width-(UIScreen.main.bounds.size.width/3-10), y: 0, width: UIScreen.main.bounds.size.width/3-10, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale);
            }
            else if (nativeAdView.nativeType == NATIVE_SMALLIMAGE){
                (nativeAdView.title == nil) ? (nativeAdView.originalityView.frame = CGRect.init(x: 0, y:0, width: UIScreen.main.bounds.size.width, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale+20)) : (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: (nativeAdView.title?.frame.size.height)!+10, width: UIScreen.main.bounds.size.width, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale+20));
                nativeAdView.originalityView.firstImageView.frame = CGRect.init(x: 0, y: 10, width: UIScreen.main.bounds.size.width/3-10, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale);
                nativeAdView.originalityView.originalityText.frame = CGRect.init(x: UIScreen.main.bounds.size.width/3, y: 0, width: UIScreen.main.bounds.size.width/3*2, height: (UIScreen.main.bounds.size.width/3-10)*nativeAdView.scale);
            }
            else if (nativeAdView.nativeType == NATIVE_SINGLETITLE){
                nativeAdView.originalityView.originalityText.font = UIFont.systemFont(ofSize: 20);
                nativeAdView.originalityView.originalityText.sizeToFit();
                (nativeAdView.title != nil) ? (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: (nativeAdView.title?.frame.size.height)!+10, width: UIScreen.main.bounds.size.width, height: nativeAdView.originalityView.originalityText.bounds.size.height)) : (nativeAdView.originalityView.frame = CGRect.init(x: 0, y: 0, width: UIScreen.main.bounds.size.width, height: nativeAdView.originalityView.originalityText.bounds.size.height));
                nativeAdView.originalityView.originalityText.frame = CGRect.init(x: 0, y: 0, width: nativeAdView.originalityView.originalityText.bounds.size.width, height: nativeAdView.originalityView.originalityText.bounds.size.height);
            }
            if(nativeAdView.additionalType == ADDITIONAL_FORM || nativeAdView.additionalType == ADDITIONAL_PHONE || nativeAdView.additionalType == ADDITIONAL_DOWNLOAD){
                nativeAdView.additionalView?.frame = CGRect.init(x: 0, y: nativeAdView.originalityView.frame.origin.y+nativeAdView.originalityView.frame.size.height, width: UIScreen.main.bounds.size.width, height: 30);
                nativeAdView.additionalView?.additionalText.frame = CGRect.init(x: 0, y: 0, width: 200, height: 30);
                nativeAdView.additionalView?.additionalButton.sizeThatFits(CGSize.init(width: 60, height: 25));
                nativeAdView.additionalView?.additionalButton.frame = CGRect.init(x: UIScreen.main.bounds.size.width-60-5, y: 2.5, width: 60, height: 25);
            }
            (nativeAdView.additionalType == ADDITIONAL_FORM || nativeAdView.additionalType == ADDITIONAL_PHONE || nativeAdView.additionalType == ADDITIONAL_DOWNLOAD) ? (nativeAdView.logoView.frame = CGRect.init(x: 0, y: (nativeAdView.additionalView?.frame.origin.y)!+(nativeAdView.additionalView?.frame.size.height)!, width: UIScreen.main.bounds.size.width, height: 20)) : (nativeAdView.logoView.frame = CGRect.init(x: 0, y: nativeAdView.originalityView.frame.origin.y+nativeAdView.originalityView.frame.size.height, width: UIScreen.main.bounds.size.width, height: 20));
            nativeAdView.logoView.adTag.sizeToFit();
            nativeAdView.logoView.adTag.frame = CGRect.init(x: 5, y: (20-nativeAdView.logoView.adTag.frame.size.height)/2, width: nativeAdView.logoView.adTag.frame.size.width, height: nativeAdView.logoView.adTag.frame.size.height);
            nativeAdView.logoView.adSource.sizeToFit();
            nativeAdView.logoView.adSource.frame = CGRect.init(x: 5+nativeAdView.logoView.adTag.frame.size.width+5, y: (20-nativeAdView.logoView.adSource.frame.size.height)/2, width: nativeAdView.logoView.adSource.frame.size.width, height: nativeAdView.logoView.adSource.frame.size.height);
            nativeAdView.logoView.closeButton.frame = CGRect.init(x: UIScreen.main.bounds.size.width-20-10, y: 0, width: 20, height: 20);
            data.height = nativeAdView.logoView.frame.origin.y+nativeAdView.logoView.frame.size.height;
            data.nativeAdView = nativeAdView;
            self._sourceArray.insert(data, at: self._indexRow+1);
            self.tableView.reloadData();
        }
 */
    }
    //关闭广告
    func didClose(_ nativeAdView: DeepleaperNativeAdView) {
        for i in 0..<_sourceArray.count{
            let data = _sourceArray[i];
            if(data.nativeAdView?.isClosed == true){
                _sourceArray.remove(at: i);
                DispatchQueue.main.async {
                    self.tableView.deleteRows(at: [IndexPath.init(row: i, section: 0)], with: .fade);
                };
                break;
            }
        }
    }
    //请求失败
    func didFailLoadNativeAdView(_ error: Error?) {
        print(error);
    }
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    /*
     // MARK: - Navigation
     
     // In a storyboard-based application, you will often want to do a little preparation before navigation
     override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
     // Get the new view controller using segue.destinationViewController.
     // Pass the selected object to the new view controller.
     }
     */
    
}
